clippy_lints/src/recursive_format_impl.rs

```diff
@@ -0,0 +1,269 @@
+                    check_to_string_in_display(cx, expr, self_hir_id);
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Display);
+                },
+                Some(ImplTrait::Debug) => {
+                    check_self_in_format_args(cx, expr, self_hir_id, ImplTrait::Debug);
+                },
+                None => {},
+            }
+        }
+    }
+}
+
+fn check_to_string_in_display(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId) {
+    if_chain! {
+        // Get the hir_id of the object we are calling the method on
+        if let ExprKind::MethodCall(path, _, [ref self_arg, ..], _) = expr.kind;
+        // Is the method to_string() ?
+        if path.ident.name == sym!(to_string);
+        // Is the method a part of the ToString trait? (i.e. not to_string() implemented
+        // separately)
+        if let Some(expr_def_id) = cx.typeck_results().type_dependent_def_id(expr.hir_id);
+        if is_diag_trait_item(cx, expr_def_id, sym::ToString);
+        // Is the method is called on self
+        if path_to_local_id(self_arg, self_hir_id);
+        then {
+            span_lint(
+                cx,
+                RECURSIVE_FORMAT_IMPL,
+                expr.span,
+                "using `to_string` in `fmt::Display` implementation might lead to infinite recursion",
+            );
+        }
+    }
+}
+
+fn check_self_in_format_args(cx: &LateContext<'_>, expr: &Expr<'_>, self_hir_id: HirId, impl_trait: ImplTrait) {
+    // Check each arg in format calls - do we ever use Display on self (directly or via deref)?
+    if_chain! {
+            if let Some(format_args) = FormatArgsExpn::parse(cx, expr);
+            let expr_expn_data = expr.span.ctxt().outer_expn_data();
+            let outermost_expn_data = outermost_expn_data(expr_expn_data);
+            if let Some(macro_def_id) = outermost_expn_data.macro_def_id;
+            if FORMAT_MACRO_PATHS
+                .iter()
+                .any(|path| match_def_path(cx, macro_def_id, path))
+                || FORMAT_MACRO_DIAG_ITEMS
+                    .iter()
+                    .any(|diag_item| cx.tcx.is_diagnostic_item(*diag_item, macro_def_id));
+            if let ExpnKind::Macro(_, _name) = outermost_expn_data.kind;
+            if let Some(args) = format_args.args();
+            then {
+                for (_i, arg) in args.iter().enumerate() {
+                    match impl_trait {
+                    // In Display, we only care about Display (it is okay to use Debug)
+                        ImplTrait::Display => {
+                    if arg.format_trait != sym::Display {
+                        continue;
+                    }},
+                    // In Debug, we only care about Debug (it is okay to use Display and ToString)
+                    ImplTrait::Debug => {
+                    if arg.format_trait != sym::Debug {
+                        continue;
+                    }},
```

Comment:
You could have `ImplTrait::name() -> Symbol` and then just do `impl_trait.name() == arg.format_trait`.